protocol/messenger_handler.go

```diff
@@ -410,10 +410,89 @@ func (m *Messenger) handleCommandMessage(state *ReceivedMessageState, message *c
 	return nil
 }
 
+func (m *Messenger) syncReceivedContactRequest(contact *Contact, state *ReceivedMessageState, chat *Chat) error {
+	contactRequestID, err := m.persistence.LatestPendingContactRequestIDForContact(contact.ID)
+	if err != nil {
+		return err
+	}
+
+	if contactRequestID != "" {
+		return nil
+	}
+
+	if chat == nil {
+		return fmt.Errorf("no chat restored during the contact synchronisation, contact.ID = %s", contact.ID)
+	}
+
+	clock, timestamp := chat.NextClockAndTimestamp(m.transport)
+	contactRequest, err := m.generateContactRequest(clock, timestamp, contact, "Please add me to your contacts", false)
+	if err != nil {
+		return err
+	}
+
+	contactRequest.ID = defaultContactRequestID(contact.ID)
+
+	// save this message
+	state.Response.AddMessage(contactRequest)
+	err = m.persistence.SaveMessages([]*common.Message{contactRequest})
+	if err != nil {
+		return err
+	}
+
+	// create AC notification
+	err = m.createIncomingContactRequestNotification(contact, state, contactRequest, true)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *Messenger) syncSentContactRequest(contact *Contact, state *ReceivedMessageState, chat *Chat) error {
```

Comment:
This and other `syncReceivedContactRequest` function are quite similar.

As far as I understood, the first part of their work is identical, except for the boolean flag passed to `m.generateContactRequest`. On the second part they differ on the call to generate a notification. At least on the first part I think it's worth an extraction refactor and parameterizing the boolean flag.

Reply:
Kk, i replaced this one `syncContactRequestForInstallationContact`